sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_event.cpp

```diff
@@ -1181,6 +1181,44 @@ ur_result_t _ur_ze_event_list_t::createAndRetainUrZeEventList(
   return UR_RESULT_SUCCESS;
 }
 
+ur_result_t _ur_ze_event_list_t::insert(_ur_ze_event_list_t &Other) {
+  if (this != &Other) {
+    // save of the previous object values
+    uint32_t PreLength = this->Length;
+    ze_event_handle_t *PreZeEventList = this->ZeEventList;
+    ur_event_handle_t *PreUrEventList = this->UrEventList;
+
+    // allocate new memory
+    uint32_t Length = PreLength + Other.Length;
+    this->ZeEventList = new ze_event_handle_t[Length];
+    this->UrEventList = new ur_event_handle_t[Length];
```

Comment:
Can we use smart pointers here?
The code looks strange as we have 2 new and 4 delete statements.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-newdelete

Reply:
This code was added to fix leaks we found with `ZE_DEBUG=4` , so I think the 4 delete statements are correct based on the absence of leaks, but agree that smart pointers is a nicer mechanism.

The existing code for the  `_ur_ze_event_list_t` object uses `new` and `delete`, so to convert to smart pointers we'd need to modify [_ur_ze_event_list_t::collectEventsForReleaseAndDestroyPiZeEventList](https://github.com/intel/llvm/blob/5b594c3c46d9974e5e3dda9561294b13aae02c5d/sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_event.cpp#L1222-L1262) which we haven't touched so far. I don't fully understand how that function works yet, but we could delve into it if using smart pointers is a valuable improvement.

Reply:
@EwanC is correct, this is something we already have in other parts of the code and we could optimize. so either we can address it here or I can follow up with a generic fix for this code and other existing in a follow-up patch.

Reply:
> @EwanC is correct, this is something we already have in other parts of the code and we could optimize. so either we can address it here or I can follow up with a generic fix for this code and other existing in a follow-up patch.

Would prefer a follow-up PR as I'm not sure that I feel confident enough in this code to touch this common functionality without causing a regression, thanks for offering @jandres742

Reply:
I will follow up.